corehq/sql_db/connections.py

```diff
@@ -36,11 +37,62 @@ def _get_db_alias_or_none(enigne_id):
         return None
 
 
+_thread_locals = local()
+
+
+READ_FROM_CITUS_STANDBYS = 'READ_FROM_CITUS_STANDBYS'
+
+
+def allow_read_from_citus_standbys():
+    return getattr(_thread_locals, READ_FROM_CITUS_STANDBYS, False)
+
+
+class read_from_citus_standbys(ContextDecorator):
+    def __enter__(self):
+        setattr(_thread_locals, READ_FROM_CITUS_STANDBYS, True)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        setattr(_thread_locals, READ_FROM_CITUS_STANDBYS, False)
+
+
+def is_citus_db(connection):
+    """
+    :param connection: either a sqlalchemy connection or a Django cursor
+    """
+    res = connection.execute("SELECT 1 FROM pg_extension WHERE extname = 'citus'")
+    if res is None:
+        res = list(connection)
+    return bool(list(res))
+
+
+def is_citus(engine, citus_list=[], non_citus_list=[]):
+    # hacking mutable default arg to memoize list of citus DBs
+    #   to avoid hitting DB. citus_list, non_citus_list
+    #   is never to be actually passed.
+    if engine.url in citus_list:
+        return True
+    if engine.url in non_citus_list:
+        return False
+    with engine.begin() as connection:
+        if is_citus_db(connection):
+            citus_list.append(engine.url)
+            return True
+        else:
+            non_citus_list.append(engine.url)
+            return False
+
+
 def create_engine(connection_string):
     # paramstyle='format' allows you to use column names that include the ')' character
     # otherwise queries will sometimes be misformated/error when formatting
     # https://github.com/zzzeek/sqlalchemy/blob/ff20903/lib/sqlalchemy/dialects/postgresql/psycopg2.py#L173
-    return sqlalchemy.create_engine(connection_string, paramstyle='format')
+    engine = sqlalchemy.create_engine(connection_string, paramstyle='format')
+
+    if allow_read_from_citus_standbys():
```

Comment:
@sravfeyn this needs to be done in the ConnectionManager since it caches the SessionHelper based on the connection url: https://github.com/dimagi/commcare-hq/blob/e7ac761cfa2ab669155142ef089fc061fcd7e863/corehq/sql_db/connections.py#L147

Reply:
https://github.com/dimagi/commcare-hq/pull/26199